library/std/src/sys/theseus/thread.rs

```diff
@@ -1,33 +1,49 @@
-use super::unsupported;
-use crate::ffi::CStr;
-use crate::io;
-use crate::num::NonZeroUsize;
-use crate::time::Duration;
+use super::{current_task, current_task_id, io_err};
+use crate::{ffi::CStr, io, num::NonZeroUsize, sys::unsupported, time::Duration};
+use libtheseus::{mem, stdio, task};
 
-pub struct Thread(!);
+pub struct Thread(task::JoinableTaskRef);
 
-pub const DEFAULT_MIN_STACK_SIZE: usize = 4096;
+pub const DEFAULT_MIN_STACK_SIZE: usize = mem::KERNEL_STACK_SIZE_IN_PAGES * mem::PAGE_SIZE;
 
 impl Thread {
     // unsafe: see thread::Builder::spawn_unchecked for safety requirements
-    pub unsafe fn new(_stack: usize, _p: Box<dyn FnOnce()>) -> io::Result<Thread> {
-        unsupported()
+    pub unsafe fn new(stack: usize, p: Box<dyn FnOnce()>) -> io::Result<Thread> {
+        let mmi_ref = mem::get_kernel_mmi_ref().ok_or_else(|| io_err("couldn't get kernel mmi"))?;
+        let stack = task::alloc_stack_by_bytes(stack, &mut mmi_ref.lock().page_table)
+            .ok_or_else(|| io_err("couldn't allocate stack"))?;
+
+        let child_task =
+            task::new_task_builder(|_| p(), ()).block().stack(stack).spawn().map_err(io_err)?;
+
+        // FIXME: We need to delete the streams when the thread exits.
```

Comment:
for now, we could print a warning about this in `exit()`. I do something similar in wasmtime to better inform users of TODOs

Reply:
which `exit()`?

Reply:
sorry i was thinking of libc's exit. brainfart, disregard.